src/Essentials/src/SecureStorage/SecureStorage.android.cs

```diff
@@ -1,395 +1,175 @@
-
-			// If >= API 23 we can use the KeyStore's symmetric key
-			if (useSymmetric && !alwaysUseAsymmetricKey)
-				return GetSymmetricKey();
-
-			// NOTE: KeyStore in < API 23 can only store asymmetric keys
-			// specifically, only RSA/ECB/PKCS1Padding
-			// So we will wrap our symmetric AES key we just generated
-			// with this and save the encrypted/wrapped key out to
-			// preferences for future use.
-			// ECB should be fine in this case as the AES key should be
-			// contained in one block.
-
-			// Get the asymmetric key pair
-			var keyPair = GetAsymmetricKeyPair();
-
-			var existingKeyStr = Preferences.Get(prefsMasterKey, null, alias);
-
-			if (!string.IsNullOrEmpty(existingKeyStr))
+			lock (locker)
 			{
-				try
-				{
-					var wrappedKey = Convert.FromBase64String(existingKeyStr);
-
-					var unwrappedKey = UnwrapKey(wrappedKey, keyPair.Private);
-					var kp = unwrappedKey.JavaCast<ISecretKey>();
-
-					return kp;
-				}
-				catch (InvalidKeyException ikEx)
+				using (var editor = GetEncryptedSharedPreferences().Edit())
 				{
-					System.Diagnostics.Debug.WriteLine($"Unable to unwrap key: Invalid Key. This may be caused by system backup or upgrades. All secure storage items will now be removed. {ikEx.Message}");
+					editor.Clear().Apply();
 				}
-				catch (IllegalBlockSizeException ibsEx)
-				{
-					System.Diagnostics.Debug.WriteLine($"Unable to unwrap key: Illegal Block Size. This may be caused by system backup or upgrades. All secure storage items will now be removed. {ibsEx.Message}");
-				}
-				catch (BadPaddingException paddingEx)
-				{
-					System.Diagnostics.Debug.WriteLine($"Unable to unwrap key: Bad Padding. This may be caused by system backup or upgrades. All secure storage items will now be removed. {paddingEx.Message}");
-				}
-				SecureStorage.RemoveAll();
 			}
-
-			var keyGenerator = KeyGenerator.GetInstance(aesAlgorithm);
-			var defSymmetricKey = keyGenerator.GenerateKey();
-
-			var newWrappedKey = WrapKey(defSymmetricKey, keyPair.Public);
-
-			Preferences.Set(prefsMasterKey, Convert.ToBase64String(newWrappedKey), alias);
-
-			return defSymmetricKey;
 		}
 
-		// API 23+ Only
-		ISecretKey GetSymmetricKey()
+		static ISharedPreferences GetEncryptedSharedPreferences()
 		{
-			Preferences.Set(useSymmetricPreferenceKey, true, SecureStorage.Alias);
+			var runtimeApiLevel = SdkInt;
+			
+			string fileName;
+			string prefsMainKey;
+			var context = Application.Context;
 
-			var existingKey = keyStore.GetKey(alias, null);
 
-			if (existingKey != null)
+			if (runtimeApiLevel >= 23)
 			{
-				var existingSecretKey = existingKey.JavaCast<ISecretKey>();
-				return existingSecretKey;
+				prefsMainKey = GetOrCreateMasterKeyApi23OrNewer();
+				fileName = "MAUI";
+			}
+			else
+			{
+				prefsMainKey = GetOrCreateMasterKeyApi22OrOlder(context);
+				fileName = "MAUI_Legacy";
```

Comment:
Is there a reason for separate files? Also, what happens if I have a API 22 phone and then upgrade? Should we maybe have an upgrade story?

Reply:
Nope, I don't think so. And yeah, the upgrade story is an interesting concept! Will have to look into that more, depending on how we decide to approach the package support for api 21/22